test/contract_tests.js

```diff
@@ -0,0 +1,498 @@
+
+// Merkle Helpers
+const { generateLeafData,
+    generateLeaf,
+    generateReportRoot,
+    generateReporterLeaf,
+    generateReportNumberLeaf,
+    generateProjectLeaf,
+    generateReportLeaf,
+    getLeafGenerator,
+    generateReportMerkleTree,
+    merkleProof,
+    generateCommitment } = require("./Utils/Merkle");
+
+// Function Argument Generators
+const { makeSubmitArgs,
+    makeAttestArgs,
+    makeDiscloseArgs,
+    makeWithdrawArgs } = require("./Utils/Generators");
+
+
+// [!] CONSTANTS KEYWORDS
+const OperatorKeccak = keccak256("OPERATOR_ROLE");
+const NativeAsset = '0x0000000000000000000000000000000000000000' // ethers.utils.getAddress('0x0') //  ETH on mainet, XDAI on paonetwork
+const AttestationDelay = 24 // 24 hours
+
+const Address_Zero = ethers.constants.AddressZero;
+const Bytes32_String = ethers.utils.formatBytes32String("hax");
+const Bytes32_Zero = ethers.utils.formatBytes32String(0);
+
+const ONE_ETHER_FORMAT = ethers.utils.parseUnits("1", "ether");
+const HALF_ETHER_FORMAT = ethers.utils.parseUnits("0.5", "ether");
+
+
+// [!] TESTING SETUP
+describe("Notary Test Cases",async function () {
+    let TestERC20Factory,
+        TestERC20;
+
+    let EscrowImpl,
+        Escrow;
+    
+    let NotaryImpl,
+        Notary;
+
+    // addresses
+    let ERC20Payer,
+        Deployer,
+        Triager,
+        Reporter1,
+        Reporter2,
+        Client;
+
+    let report1,report2
+
+    before(async function () {
+        await upgrades.silenceWarnings(); // https://forum.openzeppelin.com/t/logging-verbose-during-local-tests-with-upgradeable-contracts/4633/10
+    });
+
+    beforeEach(async () => {
+        // Mock ERC20 `TestToken` Contract - Initialization
+        TestERC20Factory = await ethers.getContractFactory("TestToken");
+        [ERC20Payer, Deployer, Triager, Reporter1,Reporter2, Client] = await ethers.getSigners();
+        TestERC20 = await TestERC20Factory.deploy(ERC20Payer.address);
+
+        // Notary Contract - Initialization
+        NotaryImpl = await ethers.getContractFactory("BugReportNotary");
+        Notary = await upgrades.deployProxy(NotaryImpl, [Deployer.address], { unsafeAllow: ['delegatecall'] });
+
+        // Escrow Contract - Initialization : Escrow contract initialize the `notary` interface with `Notary.address`
+        EscrowImpl = await ethers.getContractFactory("Escrow");
+        Escrow = await upgrades.deployProxy(EscrowImpl, [Notary.address], { unsafeAllow: ['delegatecall'] });
+
+        await TestERC20.approve(Escrow.address, ethers.utils.parseUnits("100000", "ether")); // ether format for 100000 tokens
+
+        report1 = { "id": 1, "salts": generateRandomSalt(), "paymentWalletAddress": Reporter1.address, "project": "0xbf971d4360414c84ea06783d0ea51e69035ee526" }
+        report2 = { "id": 2, "salts": generateRandomSalt(), "paymentWalletAddress": Reporter2.address, "project": "0xfbb1b73c4f0bda4f67dca266ce6ef42f520fbb98" }
```

Comment:
Ahh, the salts should be generated per-key. Not per-report. In other words, `report1.id` should have a unique salt from `report1.paymentWalletAddress` and also unique from `report2.id`

Reply:
Added.